### ISSUE TITLE
Implementation details of isometry loss

### ISSUE THREAD

**Issue by Luciennnnnnn** [isno0907/isodiff#2]
Hi, I see you set `create_graph=True` in `isometry_loss_t`, but use default value (which is False) in `isometry_loss_h`. Why is the case? I think we should set `create_graph=True` for both cases

Furthermore, why do you set `requires_grad=True` in this [line](https://github.com/isno0907/isodiff/blob/f4be9b80f90b0440dc4cc383db0eefac812d3522/utils.py#L124).

And what's the meaning of commented `torch.sqrt(detG)` in this [line](https://github.com/isno0907/isodiff/blob/f4be9b80f90b0440dc4cc383db0eefac812d3522/utils.py#L192)

**Comment by jaehoon-hahm** [comment]
Hi, thanks for your interest in our work!

I cannot remember exactly, but we were debugging and faced with higher-order derivatives. Without `create_graph=True`, we were not able to backpropagate through `isometry_loss_t`. This was not in the case `isometry_loss_h` where the loss function was written directly in the native coordinates x.

The second question also relates to this; when taking the derivative of `isometry_loss_t`, we need a derivative of the metric G. That's why we need `requires_grad=True`. But I recommend you to try out different cases by yourself.

We were experimenting with different probability measure when taking the expectation such as Dirichlet energy. Please refer to this paper: Jang, Cheongjae, et al. "Geometrically regularized autoencoders for non-Euclidean data." The Eleventh International Conference on Learning Representations. 2022.

**Comment by Luciennnnnnn** [comment]
Hi @Jaehoon-zx , thank you for your detailed response!

I have another question: Why did using `G_s` perform better than using `I` in your experiments?
Specifically, how did you conduct the experiments for Figure 3? For Figure 3.(c), is f = e_θ ∘ Π_{n-1}^{-1} or f = e_θ? If f = e_θ ∘ Π_{n-1}^{-1} was used, then it's understandable why `G = I` doesn't work well, since correct metric is `G_s`. However, if f = e_θ, why isn't `G = I` good? Isn't the Isometry Loss independent of the local coordinate system? If using Euclidean coordinates, isn't `I` the correct metric? As mentioned in Appendix B.1 of your recommend paper, it shouldn't matter. Or is this because Euclidean coordinates (n dimension) and stereographic projection coordinates(n-1 dimension) are not equivalent local coordinate systems, as Euclidean coordinates belong to the embedding space of the sphere?

In summary, my confusion is: do we necessarily need to convert to (n-1)-dimensional stereographic projection coordinates to compute the isometry loss? Because e_θ is actually defined on n-dimensional Euclidean coordinates, and to compute isometry loss in stereographic projection coordinates, we need to define f = e_θ ∘ Π_{n-1}^{-1}, which adds complexity.

**Comment by jaehoon-hahm** [comment]
Thanks, this is a good question.

For fig. 3, we used f = e_θ ∘ Π_{n-1}^{-1}, just like as we formulated the L_iso, so when the incorrect metric G=I is naively used we get sub-optimal result. Yes, the trace term is invariant under coordinate transformation, which you need rank preservation. In this case, we wanted to restrict the domain to become a n-1 dimensional hypersphere, embedded in R^n. Following the motivation, we used stereographic projection, which is not rank preserving, and hence not a coordinate transformation. (Using projection, the jacobian becomes singular and you cannot really get an inverse of Jacobian, which is required to prove the invariance of the trace term.)

I think the main difference here is what you have already noted; these are not equivalent local coordinate systems because we have used projection and restricted to a lower dimensional space. I believe when you avoid the evaluation (or the derivative) of  f = e_θ ∘ Π_{n-1}^{-1} and directly use isometry loss on R^n, this becomes a different optimization and there is no information about the spherical manifold in the loss function except for the expectation part where we sample the evaluation points from a gaussian-like pdf. This would have somewhat similar affect to that of path length regularizer, which did not explicitly take account of the spherical geometry. (Linear path penetrating the sphere would be a geodesic in this case.)

I do think that finding a way to avoid this computational complexity while achieving effective geometric regularization will certainly be an interesting future work.

**Comment by Luciennnnnnn** [comment]
I see, now I understand, thank you!

By the way, are you familiar with [Riemannian Flow Matching](https://arxiv.org/abs/2302.03660)? I've been reading their [code](https://github.com/facebookresearch/riemannian-fm/tree/main). I noticed that when implementing flow matching on spheres, they didn't use local coordinate systems(they didn't use local coordinate systems for all manifolds actually), but instead directly used Euclidean coordinates in R^n. While using local coordinates is crucial for _isodiff_, why isn't it necessary for Riemannian flow matching?

**Comment by jaehoon-hahm** [comment]
I have read the paper, but not the code. Are they also doing geometric regularization in their code? I thought they were just training modified flow matching loss function on general Riemmanian manifolds. I also want to note that ordinary path length regularization is still also a valid method even it does not specifically handles the spherical case.

**Comment by Luciennnnnnn** [comment]
@Jaehoon-zx Yes, they just training flow matching on general Riemmanian manifolds like hyperspheres

**Comment by Luciennnnnnn** [comment]
I notice that you use `r_k = math.sqrt(D_1)` in `inv_stereographic_proj_t`, and use `r = (t**2).sum(1, keepdim=True).sqrt()` in `metric_stereographic`, this is very different from equation (2), (3) of your paper. Why do you use this?

**Comment by Luciennnnnnn** [comment]
Should we pass `r` calculated in `stereographic_proj_t` to `inv_stereographic_proj_t` and `metric_stereographic`?

**Comment by jaehoon-hahm** [comment]
We used the fact that E[r] ~= sqrt(N) where N is the dimension of the space that r resides in. It seems we are calculating r for each points in sterreographic projection and metric, while taking expectation approximation in inverse stereographic projection. This could be unified and I think you could also pass the pre-computed r value to subsequent functions. This is a matter of choosing whether to calculate r for each point, or to take an approximation to its expectation.

**Comment by Luciennnnnnn** [comment]
Thank you for your help, I now understand these implementation details!

I tried to apply this loss in my setting, but couldn't see any effect. My input space is a Cartesian product of spheres (for example, a Torus is a Cartesian product of two spheres). Specifically, for an input of shape B*C*H*W, where B is batch size, C represents (C-1)-dimensional unit spheres in C-dimensional space, and H*W is the number of spheres.

I have a mapping `f` that I want to make isometric, but I found that the isometry loss doesn't seem to decrease, and test metrics like `ppl` even got worse. Could you share how the isometry loss changed during your experiments?

Here is my implementation code:
```python
from typing import Optional
import torch
import torch.autograd.functional as A


def acos_grad(z):
    return -1 / torch.sqrt(1 - z**2)


class safe_acos(torch.autograd.Function):
    @staticmethod
    def forward(ctx, input):
        ctx.save_for_backward(input)
        
        # protect ourselves from nan outputs in forward pass.
        return torch.clamp(input, min=-1 + 1e-6, max=1 - 1e-6).acos()

    @staticmethod
    def backward(ctx, grad_output):
        input, = ctx.saved_tensors
        grad_input = grad_output.clone()

        # protect ourselves from large gradients in backward pass.
        # outside of (-1 + epsilon, 1 - epsilon), gradient value is fixed constant to acos'(1-epsilon)
        epsilon = .05
        safe_input = torch.clamp(input, min=-1 + epsilon, max=1 - epsilon)

        return acos_grad(safe_input) * grad_input
    

def slerp(x1, x2, t, dims=1):
    dtype = x1.dtype
    x1 = x1.to(dtype=torch.float32)
    x2 = x2.to(dtype=torch.float32)
    x1_norm = x1 / torch.norm(x1, dim=dims, keepdim=True)
    x2_norm = x2 / torch.norm(x2, dim=dims, keepdim=True)
    omega = safe_acos.apply((x1_norm * x2_norm).sum(dim=dims, keepdim=True))
    so = torch.sin(omega)
    x = (torch.sin((1.0 - t) * omega) / so) * x1 + (torch.sin(t * omega) / so) * x2

    return x.to(dtype)


def inv_stereographic_proj_t(t, r):
    '''
    t: torch.tensor whose size = (n, D-1)
    '''
    t = t.flatten(1)
    t_norm_sq = (t ** 2).sum(1, keepdim=True)

    result = (r / (t_norm_sq + 1)) * torch.cat((2 * t, t_norm_sq - 1), dim=-1)
    
    return result


def stereographic_proj_t(x):
    '''
    x: torch.tensor whose size = (n, D)
    '''
    x = x.flatten(1)  # n, D 
    r = (x**2).sum(1, keepdim=True).sqrt()

    return (1/(r - x[:,-1:] + 1e-8)) * x[:,:-1], r # n, D-1

def metric_stereographic(t, r):
    '''
    t:      torch.tensor whose size = (n, D-1)
    out:    torch.tensor whose size = (n, D-1, D-1)
    '''
    t_norm_sq = (t ** 2).sum(1, keepdim=True) # B , 1

    G = 4 * r ** 4 / ((t_norm_sq + r ** 2) ** 2) * torch.ones_like(t, requires_grad=True) # B, D-1

    return G # B, D-1


def inv_stereographic_proj_t_spatial(t, r):
    '''
    t: torch.tensor whose size = (n, c-1, h, w)
    '''
    t_norm_sq = (t ** 2).sum(1, keepdim=True)

    result = (r / (t_norm_sq + 1)) * torch.cat((2 * t, t_norm_sq - 1), dim=1)
    
    return result

def stereographic_proj_t_spatial(x):
    '''
    x: torch.tensor whose size = (n, c, h, w)
    '''
    r = (x**2).sum(1, keepdim=True).sqrt() # n, 1, h, w

    return (1/(r - x[:,-1:] + 1e-8)) * x[:,:-1], r # n, c-1, h, w


def metric_stereographic_spatial(t, r):
    '''
    t:      torch.tensor whose size = (n, c-1, h, w)
    r:      torch.tensor whose size = (n, 1, h, w)
    out:    torch.tensor whose size = (n, c-1, h, w)
    '''
    t_norm_sq = (t ** 2).sum(1, keepdim=True) # B,1,h,w

    G = 4 * r ** 4 / ((t_norm_sq + r ** 2) ** 2) * torch.ones_like(t, requires_grad=True) # B, c-1, h, w

    return G # B, c-1, h, w

class IsometryLoss(torch.nn.Module):
    def __init__(
        self,
        isometry_batch_shrink: int = 1,
        input_domain_metric: str = "spherical",
        eta: Optional[float] = None,
    ):
        super().__init__()
        self.isometry_batch_shrink = isometry_batch_shrink
        self.input_domain_metric = input_domain_metric
        self.eta = eta

    def isometry_loss_I(self, input, func):
        bs = input.shape[0]
        u = torch.randn_like(input, device=input.device)
        Ju = A.jvp(func, input, u)[1]
        JTJu = A.vjp(func, input, Ju)[1]

        TrR = torch.sum((Ju.view(bs, -1) ** 2), dim=1).mean()
        TrR2 = torch.sum(JTJu.view(bs, -1) ** 2, dim=1).mean()
        isometry_loss = TrR2 / TrR ** 2
        return isometry_loss

    def isometry_loss_spherical(self, input, func):
        bs = input.shape[0]
        input_s, r = stereographic_proj_t(input) # B, D-1
        G = metric_stereographic(input_s, r) # B, D-1
        
        u = torch.randn_like(input_s, device=input.device)
        J_sqrt_Ginv_u = A.jvp(lambda z: func(inv_stereographic_proj_t(z, r).reshape(input.shape)), input_s, 1/torch.sqrt(G) * u, create_graph=True)[1]
        J_Ginv_u = A.jvp(lambda z: func(inv_stereographic_proj_t(z, r).reshape(input.shape)), input_s, 1/G * u, create_graph=True)[1]
        JTJ_Ginv_u = A.vjp(lambda z: func(inv_stereographic_proj_t(z, r).reshape(input.shape)), input_s, J_Ginv_u, create_graph=True)[1]

        TrR = torch.sum((J_sqrt_Ginv_u.view(bs, -1) ** 2), dim=1).mean()
        TrR2 = torch.sum(JTJ_Ginv_u.view(bs, -1) ** 2, dim=1).mean()
        isometry_loss = TrR2 / TrR ** 2
        return isometry_loss
    
    def isometry_loss_spatial_spherical(self, input, func):
        bs = input.shape[0]
        input_s, r = stereographic_proj_t_spatial(input) # B, C-1, H, W
        G = metric_stereographic_spatial(input_s, r) # B, C-1, H, W
        
        u = torch.randn_like(input_s, device=input.device) # B, C-1, H, W
        J_sqrt_Ginv_u = A.jvp(lambda z: func(inv_stereographic_proj_t_spatial(z, r).reshape(input.shape)), input_s, 1/torch.sqrt(G) * u, create_graph=True)[1]
        J_Ginv_u = A.jvp(lambda z: func(inv_stereographic_proj_t_spatial(z, r).reshape(input.shape)), input_s, 1/G * u, create_graph=True)[1]
        JTJ_Ginv_u = A.vjp(lambda z: func(inv_stereographic_proj_t_spatial(z, r).reshape(input.shape)), input_s, J_Ginv_u, create_graph=True)[1]

        TrR = torch.sum((J_sqrt_Ginv_u.view(bs, -1) ** 2), dim=1).mean()
        TrR2 = torch.sum(JTJ_Ginv_u.view(bs, -1) ** 2, dim=1).mean()
        isometry_loss = TrR2 / TrR ** 2
        return isometry_loss
    
    def forward(self, input, func):
        if self.isometry_batch_shrink != 1:
            isometry_batch_size = input.shape[0] // self.isometry_batch_shrink
            input = input[:isometry_batch_size]
        
        if self.eta is not None:
            bs = input.shape[0]
            input_perm = input[torch.randperm(bs)]
            alpha = torch.rand(bs, device=input.device) * (1 + 2 * self.eta) - self.eta
            alpha = alpha.view(-1, *([1] * (input.dim() - 1)))
            
            # input_augmented = alpha * input + (1 - alpha) * input_perm
            if self.input_domain_metric == 'euclidean':
                input_augmented = torch.lerp(input, input_perm, alpha)
            elif self.input_domain_metric == 'spherical':
                input_augmented = slerp(input, input_perm, alpha, dims=(1, 2, 3))
            elif self.input_domain_metric == 'spatial_spherical':
                input_augmented = slerp(input, input_perm, alpha, dims=(1))
            else:
                raise ValueError(f"Invalid input domain metric: {self.input_domain_metric}")
            input = input_augmented

        if self.input_domain_metric == 'spherical':
            return self.isometry_loss_spherical(input, func)
        elif self.input_domain_metric == 'spatial_spherical':
            return self.isometry_loss_spatial_spherical(input, func)
        elif self.input_domain_metric == 'euclidean':
            return self.isometry_loss_I(input, func)
        else:
            raise ValueError(f"Invalid input domain metric: {self.input_domain_metric}")

```

Is this implementation when `input_domain_metric = 'spatial_spherical'` correct?

**Comment by Luciennnnnnn** [comment]
I set `isometry_batch_shrink=2` for efficiency as it is used in path regularization by styleganv2. Also I use `eta=0.2` as done by [IRVAE](https://github.com/Gabe-YHLee/IRVAE-public)

**Comment by Luciennnnnnn** [comment]
<img width="1743" alt="image" src="https://github.com/user-attachments/assets/63f49ddd-e834-49dc-8b14-92fafba74308" />
This is my isometry loss, which seems to be very unstable

**Comment by Luciennnnnnn** [comment]
I found there may have some numerical issues:

`TrR` is around 10^8, `TrR2` is around 10^15, `TrR^2 `is around 10^17, and `J_sqrt_Ginv_u`, `J_Ginv_u` contain many values in 10^2-10^3, `JTJ_Ginv_u` contain many values 10^5-10^6

**Comment by jaehoon-hahm** [comment]
Yes, these loss functions consist of pretty numerically unstable values and it could be hard to optimize. To give a comment, we used gradient accumulation to acheive very high accuracy of the gradient and used very small learning rate. We don't have the learning history at that time, but you should see a gradual decrease in the exponential moving average of your loss function. Usually ppl proportionally decreases with the loss value. 

Naively overviewing your method, you are optimizing over HW hyperspheres which could complicate your optimization problem hence leading to the hardness of it.

**Comment by Luciennnnnnn** [comment]
I use batch size `32` and learning rate `5e-5`, the weight of isometry loss is set to `1` cause I found it have relatively small magnitude. I notice that you use lower learning rate `1e-5` and isometry loss weight `1e-3`, maybe it will affect results.
